src/pdn/src/PdnGen.cc

```diff
@@ -78,131 +75,96 @@ void PdnGen::init(dbDatabase* db, Logger* logger)
+  // only search for instances once
+  std::map<std::string, std::set<dbInst*>> inst_map;
+  for (auto* connect : connects) {
+    const std::string inst_pattern = connect->getInstPattern();
+    if (inst_map.find(inst_pattern) != inst_map.end()) {
+      continue;
+    }
+    inst_map[inst_pattern] = connect->getInsts();
   }
 
-  // Do regions
-  for (auto& [region, net_regex_pairs] : *global_connect_) {
-    if (region == nullptr)
-      continue;
-    globalConnectRegion(block, region, net_regex_pairs);
+  for (auto* connect : connects) {
+    globalConnect(connect, inst_map[connect->getInstPattern()]);
   }
 }
 
-void PdnGen::globalConnectRegion(dbBlock* block,
-                                 dbBox* region,
-                                 std::shared_ptr<netRegexPairs> global_connect)
+void PdnGen::globalConnect(dbGlobalConnect* global_connect)
 {
-  for (auto& [net, regex_pairs] : *global_connect) {
-    for (auto& [inst_regex, pin_regex] : *regex_pairs) {
-      globalConnectRegion(block, region, inst_regex, pin_regex, net);
-    }
+  if (global_connect == nullptr) {
+    return;
   }
+
+  globalConnect(global_connect, global_connect->getInsts());
 }
 
-void PdnGen::globalConnectRegion(dbBlock* block,
-                                 dbBox* region,
-                                 std::shared_ptr<regex>& instPattern,
-                                 std::shared_ptr<regex>& pinPattern,
-                                 dbNet* net)
+void PdnGen::globalConnect(dbGlobalConnect* global_connect, const std::set<dbInst*> insts)
 {
-  if (net == nullptr) {
-    logger_->warn(PDN, 60, "Unable to add invalid net.");
+  if (insts.empty()) {
     return;
   }
 
+  auto* net = global_connect->getNet();
   if (net->isDoNotTouch()) {
     return;
   }
 
-  std::vector<dbInst*> insts;
-  findInstsInArea(block, region, instPattern, insts);
-
-  if (insts.empty()) {
+  const auto mterm_map = global_connect->getMTerms();
+  if (mterm_map.empty()) {
     return;
   }
 
-  std::map<dbMaster*, std::vector<dbMTerm*>> masterpins;
-  buildMasterPinMatchingMap(block, pinPattern, masterpins);
-
   for (dbInst* inst : insts) {
-    dbMaster* master = inst->getMaster();
-
-    auto masterpin = masterpins.find(master);
-    if (masterpin != masterpins.end()) {
-      std::vector<dbMTerm*>* mterms = &masterpin->second;
-
-      for (dbMTerm* mterm : *mterms) {
-        auto* iterm = inst->getITerm(mterm);
-        auto* current_net = iterm->getNet();
-        if (current_net != nullptr && current_net->isDoNotTouch()) {
-          continue;
-        }
-        iterm->connect(net);
-        iterm->setSpecial();
-      }
-    }
-  }
-}
-
-void PdnGen::findInstsInArea(dbBlock* block,
-                             dbBox* region,
-                             std::shared_ptr<regex>& instPattern,
-                             std::vector<dbInst*>& insts)
-{
-  for (dbInst* inst : block->getInsts()) {
     if (inst->isDoNotTouch()) {
       continue;
     }
```

Comment:
In that case we would also need to check the pins on the inst to ensure the match is fully correct and probably no need to generate a warning if it's already connected correctly. This might generate a huge volume of warnings.

Reply:
Why do you think it might generate "a huge volume of warnings"?  I would think it rare that someone would have a global connect that conflicts with a dont touch instance.  We could limit ourselves to one warning per instance/global_connect.